nemo/collections/nlp/modules/common/megatron/adapters/mcore_mixins.py

```diff
@@ -0,0 +1,106 @@
+# Copyright (c) 2023, NVIDIA CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import torch
+from megatron.core.models.gpt.gpt_embedding import GPTEmbedding
+from megatron.core.transformer.attention import SelfAttention
+
+from nemo.collections.nlp.modules.common.megatron.adapters.parallel_adapters import (
+    AdapterName,
+    LoraKQVAdapterConfig,
+    PromptEncoderAdapterConfig,
+)
+from nemo.core import adapter_mixins
+
+
+def swap_mcore_mixin(module, mcore_mixin):
+    """
+    Casts module to mcore_mixin and register corresponding adapters.
+    """
+    module.__class__ = mcore_mixin
+    module.mcore_register_adapters()
+
+
+class MCoreAdapterModuleMixin(adapter_mixins.AdapterModuleMixin):
+    def mcore_register_adapters(self):
+        raise NotImplementedError("Mcore mixins should implement setup_adapters on a subclass of MyBase")
+
+
+class MCoreSelfAttentionMixin(SelfAttention, MCoreAdapterModuleMixin):
+    def mcore_register_adapters(self):
+        self.set_accepted_adapter_types([LoraKQVAdapterConfig._target_])  # only self attn (packed qkv) for now
+        self.linear_qkv.return_layernorm_output = True  # need layernorm output for lora mlp
```

Comment:
Should this be private? Could we add a docstring?

Reply:
done.